hashes/src/sha256.rs

```diff
@@ -74,41 +74,60 @@ impl HashEngine {
         }
     }
 
-    /// Create a new [`HashEngine`] from a [`Midstate`].
+    /// Creates a new [`HashEngine`] from a [`Midstate`].
     ///
-    /// # Panics
-    ///
-    /// If `length` is not a multiple of the block size.
-    pub fn from_midstate(midstate: Midstate, length: usize) -> HashEngine {
-        assert!(length % BLOCK_SIZE == 0, "length is no multiple of the block size");
-
+    /// This function should not be used directly unless you really know what you're doing i.e., you
+    /// probably should be a professional cryptographer.
+    pub fn from_midstate(midstate: Midstate) -> HashEngine {
         let mut ret = [0; 8];
-        for (ret_val, midstate_bytes) in ret.iter_mut().zip(midstate[..].chunks_exact(4)) {
+        for (ret_val, midstate_bytes) in ret.iter_mut().zip(midstate.as_ref().chunks_exact(4)) {
             *ret_val = u32::from_be_bytes(midstate_bytes.try_into().expect("4 byte slice"));
         }
 
-        HashEngine { buffer: [0; BLOCK_SIZE], h: ret, length }
+        HashEngine { buffer: [0; BLOCK_SIZE], h: ret, length: midstate.length }
     }
 
+    /// Returns `true` if the midstate can be extracted from this engine.
+    ///
+    /// The midstate can only be extracted if the number of bytes input into
+    /// the hash engine is a multiple of 64. See caveat on [`Self::midstate`].
+    pub fn can_extract_midstate(&self) -> bool { self.length % 64 == 0  }
 
-    /// Outputs the midstate of the hash engine. This function should not be
-    /// used directly unless you really know what you're doing.
+    /// Outputs the midstate of the hash engine.
+    ///
+    /// This function should not be used directly unless you really know what you're doing i.e., you
+    /// probably should be a professional cryptographer.
+    pub fn midstate(&self) -> Result<Midstate, MidstateError> {
+        if !self.can_extract_midstate() {
+            return Err(MidstateError {
+                invalid_length: self.length
+            })
+        }
+        Ok(self.midstate_())
+    }
+
+    /// Outputs the midstate of the hash engine.
     #[cfg(not(hashes_fuzz))]
-    pub fn midstate(&self) -> Midstate {
+    fn midstate_(&self) -> Midstate {
```

Comment:
A better name would be appreciated. E.g. `midstate_unchecked`.

Reply:
Maybe `midstate_inner` or `midstate_internal`.

Reply:
The method doesn't perform `% 64` check, so I thin `_unchecked` is appropriate.

Reply:
Damn, I saw this style in another crate and hoped I'd get it past you :) Will fix.

Reply:
If it's our crate we should fix it. If it's a foreign crate it's a tiny little red flag. But really a very little one.

Reply:
Note please, at first I had `midstate_internal` but PR now uses `midstate_unchecked`, I think it helps at internal call sites.